nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/adapters/HiveMqV5ClientAdapter.java

```diff
@@ -0,0 +1,148 @@
+                    .applyWillPublish();
+        }
+
+        final String username = connectionProperties.getUsername();
+        final String password = connectionProperties.getPassword();
+        if (username != null && password != null) {
+            connectBuilder.simpleAuth()
+                    .username(connectionProperties.getUsername())
+                    .password(password.getBytes(StandardCharsets.UTF_8))
+                    .applySimpleAuth();
+        }
+
+        final Mqtt5Connect mqtt5Connect = connectBuilder.build();
+        mqtt5BlockingClient.connect(mqtt5Connect);
+    }
+
+    @Override
+    public void disconnect(long disconnectTimeout) {
+        logger.debug("Disconnecting client");
+        // Currently it is not possible to set timeout for disconnect with HiveMQ Client. (Only connect timeout exists.)
+        mqtt5BlockingClient.disconnect();
+    }
+
+    @Override
+    public void close() {
+        // there is no paho's close equivalent in hivemq client
+    }
+
+    @Override
+    public void publish(String topic, StandardMqttMessage message) {
+        logger.debug("Publishing message to {} with QoS: {}", topic, message.getQos());
+
+        mqtt5BlockingClient.publishWith()
+                .topic(topic)
+                .payload(message.getPayload())
+                .retain(message.isRetained())
+                .qos(Objects.requireNonNull(MqttQos.fromCode(message.getQos())))
+                .send();
+    }
+
+    @Override
+    public void subscribe(String topicFilter, int qos) {
+        Objects.requireNonNull(callback, "callback should be set");
+
+        logger.debug("Subscribing to {} with QoS: {}", topicFilter, qos);
+
+        CompletableFuture<Mqtt5SubAck> ack = mqtt5BlockingClient.toAsync().subscribeWith()
+                .topicFilter(topicFilter)
+                .qos(Objects.requireNonNull(MqttQos.fromCode(qos)))
+                .callback(mqtt5Publish -> {
+                    final ReceivedMqttMessage receivedMessage = new ReceivedMqttMessage(
+                            mqtt5Publish.getPayloadAsBytes(),
+                            mqtt5Publish.getQos().getCode(),
+                            mqtt5Publish.isRetain(),
+                            mqtt5Publish.getTopic().toString());
+                    callback.messageArrived(receivedMessage);
+                })
+                .send();
+
+        // Setting "listener" callback is only possible with async client, though sending subscribe message
+        // should happen in a blocking way to make sure the processor is blocked until ack is not arrived.
+        ack.whenComplete((mqtt5SubAck, throwable) -> {
+            logger.debug("Received mqtt5 subscribe ack: {}", mqtt5SubAck.toString());
+
+            if (throwable != null) {
+                throw new MqttException("An error has occurred during sending subscribe message to broker", throwable);
+            }
+        });
```

Comment:
`whenComplete()` registers an async callback, it does not block. The blocking way would be to use `ack.get()`.

`whenComplete()` can also work but in this case:
- in case of error (`throwable != null`), `logger.error()` must be used instead of throwing an exception in order to put the error message on the processor's bulletin
- please fix the comment because this approach does not block

Reply:
It confused me that there is also whenCompleteAsync and because of that I thought whenComplete blocks. Thanks for noticing it! Changed.